### PR TITLE
Add framework for multi-node requests

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
Super rough, initial implementation that compiles.

**Comment by gavin-norman-sociomantic** [comment]
Based on #199.

**Comment by gavin-norman-sociomantic** [comment]
Updated: tidied up. No longer WIP. Still blocked, as it's based on #199.

**Comment by gavin-norman-sociomantic** [comment]
Dependency merged. Unblocked.

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 1]
```diff
@@ -20,8 +20,15 @@ module swarm.neo.client.mixins.RequestCore;
 
 public enum RequestType
 {
+    // A request that operates on one node at a time. The node currently being
+    // operated on may change over the request's lifetime.
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: DDOC comments should be used instead.

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 2]
```diff
@@ -20,8 +20,15 @@ module swarm.neo.client.mixins.RequestCore;
...
     SingleNode,
+
+    // A request that operates on one node at a time. If the request fails on
+    // one node, it is repeated on the next in a round-robin sequence.
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: Does the request need to fail to be repeated on the next in a round-robin sequence? Is there are request that always tries on the next node in a round-robin sequence (like `Produce`)?

File: src/swarm/neo/client/mixins/RequestCore.d (hunk 1 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: Oh yeah, I missed that.

File: src/swarm/neo/client/mixins/RequestCore.d (hunk 2 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: I think currently not, but I can make the comment more generic.

**Comment by gavin-norman-sociomantic** [comment]
Updated with suggestions re: comments.

**Comment by gavin-norman-sociomantic** [comment]
I just noticed there's loads of TODOs in here. Will fix and update.

**Comment by gavin-norman-sociomantic** [comment]
Updated: fixed doc TODOs.

**Comment by nemanja-boric-sociomantic** [comment]
Let's merge #211 first and rebase this, so we can confirm that the tests are passing.

**Comment by nemanja-boric-sociomantic** [comment]
It would be helpful in this case to run `git rebase -i -x 'git commit --amend --date=$(date +%s) --no-edit' upstream/v4.x.x` to get the commit order right in GitHub as well.

**Comment by gavin-norman-sociomantic** [comment]
Rebased on v4.x.x.

**Comment by nemanja-boric-sociomantic** [comment]
That's better, the tests are red ;-)

**Comment by gavin-norman-sociomantic** [comment]
Fixed D2 error.

File: src/swarm/neo/client/RequestSet.d [hunk 3]
```diff
@@ -263,6 +265,8 @@ public final class RequestSet: IRequestSet
             suspended fiber, waiting for I/O to complete, or may have finished.
 
             Params:
+                RequestContext = type of request context. Must be a type that
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/RequestSet.d:305]: I'll not insist to put this in a separate commit.

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/RequestSet.d:305]: Surely it counts as "Add missing method doc"?

File: integrationtest/neo/client/Client.d [hunk 4]
```diff
@@ -90,6 +93,32 @@ public class Client
             this.assign!(Internals.Put)(params);
         }
 
+        /***********************************************************************
+
+            Assigns a DoublePut request, instructing the node to associate the
+            specified value and key.
```
Comment (nemanja-boric-sociomantic) [inline integrationtest/neo/client/Client.d]: Perhaps document that this puts the `(key, value)` pair into two nodes.

File: integrationtest/neo/client/Client.d [hunk 5]
```diff
@@ -276,6 +305,63 @@ public class Client
             return succeeded;
         }
 
+        /***********************************************************************
+
+            Assigns a DoublePut request, instructing two nodes to associate the
+            specified value and key. The calling Task is blocked until the
+            request finishes.
```
Comment (nemanja-boric-sociomantic) [inline integrationtest/neo/client/Client.d:312]: (re: previous comment) - just copy the docs from here.

File: integrationtest/neo/client/request/internal/DoublePut.d [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        {
+            None,
+            Failure,    // Default value; unknown error (presumably in client)
+            Error,      // Node or I/O error
+            Success     // Put record
+        }
+
+        /// The way in which the request ended. Used by the finished notifier to
+        /// decide what kind of notification (if any) to send to the user.
+// TODO
```
Comment (nemanja-boric-sociomantic) [inline integrationtest/neo/client/request/internal/DoublePut.d]: TODO

File: integrationtest/neo/client/request/internal/DoublePut.d [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        // the nodes' addresses/ports.
+        ushort[2] ports = [10_000, 10_010];
+        AddrPort node;
+        node.setAddress("127.0.0.1");
+        node.port = ports[node_idx];
+
+        // If the selected node is the first, start the request on the second.
+        if ( node_idx == 0 )
+            start_request_on_new_conn();
+// TODO: maybe it would be easier to specify the connection at this stage
```
Comment (nemanja-boric-sociomantic) [inline integrationtest/neo/client/request/internal/DoublePut.d]: more TODOs

File: integrationtest/neo/client/Client.d (hunk 5 above)
Comment (gavin-norman-sociomantic) [reply integrationtest/neo/client/Client.d:312]: Mm do you mean copy "Assigns a DoublePut request, instructing two nodes to associate the specified value and key." to line 98? The doc is identical.

File: integrationtest/neo/client/request/internal/DoublePut.d (hunk 7 above)
Comment (gavin-norman-sociomantic) [reply integrationtest/neo/client/request/internal/DoublePut.d]: Updated.

File: integrationtest/neo/client/Client.d (hunk 5 above)
Comment (nemanja-boric-sociomantic) [reply integrationtest/neo/client/Client.d:312]: Yes, to copy that, but no, the doc is not identical as far as I can see (the doc at line 98 speaks about singular node).

Comment (gavin-norman-sociomantic) [reply integrationtest/neo/client/Client.d:312]: Ahh sorry, yes, you're right.

Comment (gavin-norman-sociomantic) [reply integrationtest/neo/client/Client.d:312]: Updated.

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Pull request merged**